pkg/antctl/raw/multicluster/join.go

```diff
@@ -0,0 +1,318 @@
+	command.Flags().StringVarP(&joinOpts.MemberClusterOpt.Namespace, "namespace", "n", "", "Namespace of the member cluster")
+	command.Flags().StringVarP(&joinOpts.MemberClusterOpt.ClusterID, "clusterid", "", "", "Cluster ID of the member cluster")
+	command.Flags().StringVarP(&joinOpts.ClusterSetID, "clusterset", "", "", "ClusterSet ID")
+	command.Flags().StringVarP(&joinOpts.SecretFile, "secret-file", "", "", "Secret file to leader cluster. If both Secret and Secret file are specified, the Secret file will be used.")
+	command.Flags().StringVarP(&joinOpts.ConfigFile, "config-file", "f", "", "Config file that defines all config options. If both command line options and config file are specified, "+
+		"the arguments in config file will be used.")
+
+	return command
+}
+
+func joinRunE(cmd *cobra.Command, args []string) error {
+	var err error
+	if err = joinOpts.validateAndComplete(); err != nil {
+		return err
+	}
+	k8sClient, err := common.NewClient(cmd)
+
+	memberClusterNamespace := joinOpts.MemberClusterOpt.Namespace
+	memberClusterID := joinOpts.MemberClusterOpt.ClusterID
+	memberClusterSet := joinOpts.ClusterSetID
+	createdRes := []map[string]interface{}{}
+	defer func() {
+		if err != nil {
+			fmt.Fprintf(cmd.OutOrStdout(), "Failed to join the ClusterSet. Deleting the created resources.\n")
+			if err := common.Rollback(cmd, k8sClient, createdRes); err != nil {
+				fmt.Fprintf(cmd.OutOrStdout(), "Failed to rollback: %v\n", err)
+			}
+		}
+	}()
+	if joinOpts.SecretFile != "" {
+		if err := importSecret(cmd, k8sClient); err != nil {
+			fmt.Fprintf(cmd.OutOrStdout(), "Failed to import secret: %v\n", err)
+			return err
+		}
+	}
+	err = common.CreateClusterClaim(cmd, k8sClient, memberClusterNamespace, memberClusterSet, memberClusterID, &createdRes)
+	if err != nil {
+		return err
+	}
+	err = common.CreateClusterSet(cmd, k8sClient, memberClusterNamespace, memberClusterSet, joinOpts.LeaderClusterOpt.APIServer, joinOpts.LeaderClusterOpt.Secret,
+		joinOpts.MemberClusterOpt.ClusterID, joinOpts.LeaderClusterOpt.ClusterID, joinOpts.LeaderClusterOpt.Namespace, &createdRes)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(cmd.OutOrStdout(), "Waiting for member cluster ready\n")
+	if err = waitForMemberClusterReady(cmd, k8sClient); err != nil {
+		fmt.Fprintf(cmd.OutOrStdout(), "Failed to wait for member cluster ready: %v\n", err)
+		return err
+	}
+	fmt.Fprintf(cmd.OutOrStdout(), "Member cluster joined successfully\n")
+
+	return nil
+}
+
+func importSecret(cmd *cobra.Command, k8sClient client.Client) error {
+	content, err := ioutil.ReadFile(joinOpts.SecretFile)
+	if err != nil {
+		return err
+	}
+	decoder := yamlutil.NewYAMLOrJSONDecoder(bytes.NewReader(content), 100)
+	var secret v1.Secret
+	if err = decoder.Decode(&secret); err != nil {
+		return err
+	}
+
+	if err := k8sClient.Create(context.TODO(), &secret); err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
We should create the resources in the provided Namespace. Do not assume the input YAML includes the Namespace.

Reply:
Add a validator to check the Namespaces now.